### PR TITLE
Add endpoint to allow creation of admin accounts

### PR DESCRIPTION
**Pull request by Gabriel-Rockson**
- feat: Move the roles column from Admin entity to the upper User entity
- test: Add tests for PermissionServiceImpl
- feat: Add AdminDTO and CreateAdminDTO
- feat: Add UserPrinciple to return user details to the auth manager
- feat: Implement the UserDetailsService and loadUserByUsername
- feat: Add security filter chain
- feat: Add password generator to generate random password for accounts
- feat: Add new constructor
- feat: Set the role for admin in the service.
- feat: Add a new constructor.
- feat: Add mapper for converting data from dto to entity and vice versa
- feat: Add Controller for adding admin accounts
- feat: Add global exception handler

### CODE REVIEW

File: src/main/java/com/amalitech/mentorshiptrackr/controllers/AdminController.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+@RequiredArgsConstructor
+public class AdminController {
+
+    private final UserService userService;
+    private final AdminMapper adminMapper;
+
+    @PostMapping()
+    @ResponseStatus(HttpStatus.CREATED)
+    public AdminDTO addNewAdmin(@RequestBody @Valid CreateAdminDTO adminDTO) {
+        Admin admin = (Admin) userService.addNewAdminAccount(adminMapper.toEntity(adminDTO));
```
Comment (daniel-bontii) [inline src/main/java/com/amalitech/mentorshiptrackr/controllers/AdminController.java:24]: Do this logic at the service layer

File: src/main/java/com/amalitech/mentorshiptrackr/dto/AdminMapper.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+import com.amalitech.mentorshiptrackr.models.Admin;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public class AdminMapper {
+
+    public AdminDTO toDTO(Admin entity) {
```
Comment (daniel-bontii) [inline src/main/java/com/amalitech/mentorshiptrackr/dto/AdminMapper.java:11]: You can easily convert this in one line of code using the object mapper
(com.fasterxml.jackson.databind.ObjectMapper)
Look into the convertValue method

File: src/main/java/com/amalitech/mentorshiptrackr/utils/PasswordGenerator.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
+package com.amalitech.mentorshiptrackr.utils;
+
+import org.springframework.stereotype.Component;
+
+import java.security.SecureRandom;
+
+@Component
+public abstract class PasswordGenerator {
```
Comment (daniel-bontii) [inline src/main/java/com/amalitech/mentorshiptrackr/utils/PasswordGenerator.java:8]: This is cool. But there are packages that can help you here

File: src/test/java/com/amalitech/mentorshiptrackr/services/PermissionServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        // given: setup inputs and targets
+        Permission permission = Permission.builder()
+                .id(UUID.randomUUID())
+                .name("Read admin list")
+                .description("View list of all admins in system.").build();
+
+        given(permissionRepository.save(permission)).willReturn(permission);
+
+        // when: act on the target
+        Permission new_permission = permissionService.addNewPermission(permission);
```
Comment (daniel-bontii) [inline src/test/java/com/amalitech/mentorshiptrackr/services/PermissionServiceTest.java:75]: new_permission is python convention 😉 newPermission is more like java.

File: src/main/java/com/amalitech/mentorshiptrackr/dto/AdminMapper.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        dto.setEmail(entity.getEmail());
+        dto.setCreatedAt(entity.getAuditData().getCreatedAt());
+        dto.setUpdatedAt(entity.getAuditData().getUpdatedAt());
+        dto.setRole(entity.getRole().getName());
+
+        return dto;
+    }
+
+    public Admin toEntity(CreateAdminDTO dto) {
+        Admin entity = new Admin();
```
Comment (daniel-bontii) [inline src/main/java/com/amalitech/mentorshiptrackr/dto/AdminMapper.java:26]: You can easily convert this in one line of code using the object mapper
(com.fasterxml.jackson.databind.ObjectMapper)
Look into the convertValue method

File: src/main/java/com/amalitech/mentorshiptrackr/utils/PasswordGenerator.java (hunk 3 above)
Comment (Gabriel-Rockson) [reply src/main/java/com/amalitech/mentorshiptrackr/utils/PasswordGenerator.java:8]: Alrigh

File: src/main/java/com/amalitech/mentorshiptrackr/controllers/AdminController.java (hunk 1 above)
Comment (Gabriel-Rockson) [reply src/main/java/com/amalitech/mentorshiptrackr/controllers/AdminController.java:24]: I don't get this please.

**Pull request closed without merging**